### ISSUE TITLE
Correct display of possible tokens / productions in parser

### ISSUE THREAD

**Issue by joshuabach** [comprakt/comprakt#44]
In many `UnexpectedToken` errors the string after `found {} expected ...` is not an exhaustive description of all possible tokens

**Comment by reiner-dolp** [comment]
We should look into two things here:
1. Multiple possibilities are checked with a chain of `if-else` Blocks. Only one of the possibilities is listed in the error message. We could just enumerate all possibilities.
2. We want some mechanism that allows us to mark errors with a superstructure. E.g.
change error messages depending on super structure
```
error: expected an identifier, found int
   | 
 3 |         return new int();
   |                    ^^^

Compilation aborted due to an error
```
Could be a
```
error: invalid new object expression. You cannot use `new` on a primitive type
   | 
 3 |         return new int();
   |                    ^^^ unexpected primitive type `int`
   |                ^^^^^^^^^ in this new object expression

```
or 
```
error: invalid new object expression. You cannot use `new` on a primitive type
   | 
 3 |         return new int();
   |                    ^^^ expected an identifier, found int

```
(This is not about formatting, but we may want to expand the span or change the error message depending on the structure around an expression.

What we want is exactly is the behaviour of error chaining possible through `?` so what we could do is revert to the variant of spanned! with a closure, the code inside the spanned!-closure will parse the infra structure, and the code before and after the spanned!-macro defines the superstructure. The code could then check for a specific error instance, e.g. unexpected keyword during parsing of new-object-expression and then alter the error message. As going deeper than one level might be to much (error chaining over more than one level and therefore two or more stacked spans), we could probably get away with something like adding a Variant `UnexpectedTokenInSuperstructure`
```rust
#[derive(Debug, Clone, Fail)]
pub enum SyntaxError {
    #[fail(display = "expected {}, found {}", expected, actual)]
    UnexpectedToken {
        actual: String,
        expected: String,
    },
    #[fail(display = "unexpected end of file")]
    UnexpectedEOF,
     #[fail(display = "{}", context)]
     UnexpectedTokenInSuperstructure {
        actual: String,
        expected: String,
        context: String
    },
}

```

**Comment by reiner-dolp** [comment]
Another example: Current rendering is
```
rror: expected identifier String, found int
   | 
 3 |     public static void main(int[] args) {
   |                             ^^^

Compilation aborted due to an error
```
better would be
```
error: invalid main method. It must be declared as `public static void main(String[] args)`
   | 
 3 |     public static void main(int[] args) {
   |                             ^^^ expected `String` instead of `int`

Compilation aborted due to an error

**Comment by joshuabach** [comment]
Closed by #95
